src/components/chats/ChatList/hooks/useGetChatElement.ts

```diff
@@ -0,0 +1,130 @@
+import useWrapCallbackInRef from '@/hooks/useWrapCallbackInRef'
+import { useIsAnyQueriesLoading } from '@/subsocial-query'
+import { generateManuallyTriggeredPromise } from '@/utils/promise'
+import { PostData } from '@subsocial/api/types'
+import { UseQueryResult } from '@tanstack/react-query'
+import { useEffect, useMemo, useRef } from 'react'
+import { getChatItemId } from '../../helpers'
+
+export default function useGetChatElement(
+  commentIds: string[],
+  commentsQuery: UseQueryResult<PostData | null, unknown>[],
+  loadedCommentsQuery: UseQueryResult<PostData | null, unknown>[],
+  loadMore: () => void
+) {
+  const waitAllCommentsLoaded = useWaitCommentsLoading(commentsQuery)
+
+  const waitAllCommentsLoadedRef = useWrapCallbackInRef(waitAllCommentsLoaded)
+  const commentIdsRef = useWrapCallbackInRef(commentIds)
+
+  const promiseRef = useRef<{
+    resolvers: Map<string, VoidFunction[]>
+    waitingCommentIds: Set<string>
+  }>({ resolvers: new Map(), waitingCommentIds: new Set() })
+
+  const loadedCommentIds = useMemo(() => {
+    const ids: string[] = []
+    loadedCommentsQuery.forEach((comment) => {
+      if (comment.data?.id) {
+        ids.push(comment.data.id)
+      }
+    })
+    return ids
+  }, [loadedCommentsQuery])
+
+  const awaitableLoadMore = useAwaitableLoadMore(
+    loadMore,
+    loadedCommentIds.length
+  )
+
+  useEffect(() => {
+    const { waitingCommentIds } = promiseRef.current
+    waitingCommentIds.forEach((commentId) => {
+      if (loadedCommentIds.includes(commentId)) {
+        const resolvers = promiseRef.current.resolvers.get(commentId)
+        resolvers?.forEach((resolve) => resolve())
+        promiseRef.current.resolvers.delete(commentId)
+        waitingCommentIds.delete(commentId)
+      }
+    })
+  }, [loadedCommentIds])
+
+  const loadMoreUntilCommentIsLoaded = async (commentId: string) => {
+    const isCommentIdIncluded = commentIdsRef.current.includes(commentId)
+    if (isCommentIdIncluded) return
+    await awaitableLoadMore()
+    await loadMoreUntilCommentIsLoaded(commentId)
+  }
+
+  const getChatElement = async (commentId: string) => {
+    const { getPromise, getResolver } = generateManuallyTriggeredPromise()
+    const elementId = getChatItemId(commentId)
+    const element = document.getElementById(elementId)
+    if (element) return element
+
+    const isCommentIdIncluded = commentIds.includes(commentId)
+    if (!isCommentIdIncluded) {
+      await loadMoreUntilCommentIsLoaded(commentId)
+    }
+    const { resolvers, waitingCommentIds } = promiseRef.current
+    if (!resolvers.get(commentId)) {
+      resolvers.set(commentId, [])
+    }
+    resolvers.get(commentId)?.push(getResolver())
+    waitingCommentIds.add(commentId)
+    await getPromise()
+    await new Promise<void>((resolve) =>
+      setTimeout(async () => {
+        await waitAllCommentsLoadedRef.current()
+        resolve()
+      })
```

Comment:
Why do you use lot of refs here?

This Promise looks dirty. Why do you use setTimeout, is it for unblocked call?

Reply:
ah I forgot to remove it, removed in [ead3178](https://github.com/dappforce/grillchat/pull/38/commits/ead3178d7255c04ba738297ebf0eedb36f879116)

there's a bug before, so I tried using it and it also don't change anything, but I forgot to remove it